modules/urlsigning-service-impl/src/main/java/org/opencastproject/security/urlsigning/provider/impl/AbstractUrlSigningProvider.java

```diff
@@ -150,121 +109,113 @@ public void addKeyEntry(String keyId, String url, String key, String organizatio
-      }
-    }
-    return null;
+  private Optional<Map.Entry<String, KeyEntry>> getKeyEntry(String baseUrl) {
+    return keys.entrySet().stream()
+            .filter(entry -> baseUrl.startsWith(entry.getValue().url))
+            .findAny();
   }
 
-  @SuppressWarnings("rawtypes")
   @Override
-  public void updated(Dictionary properties) throws ConfigurationException {
+  public void updated(Dictionary<String, ?> properties) throws ConfigurationException {
     getLogger().info("Updating {}", toString());
     if (properties == null) {
       getLogger().warn("{} is unconfigured", toString());
       return;
     }
 
-    // Clear the current set of keys
-    keys.clear();
-
-    String key = null;
-    String keyId = null;
-    String url = null;
-    String organization = null;
+    // Collect configuration in a new map so we don't partially override the old one in case of error
+    Map<String, KeyEntry> keys = new HashMap<>();
 
-    int i = 1;
-    while (true) {
-      // Create the configuration prefixes
-      key = new StringBuilder(KEY_PREFIX).append(".").append(i).toString();
-      keyId = new StringBuilder(ID_PREFIX).append(".").append(i).toString();
-      url = new StringBuilder(URL_PREFIX).append(".").append(i).toString();
-      organization = new StringBuilder(ORGANIZATION_PREFIX).append(".").append(i).toString();
-      getLogger().debug("Looking for configuration of {}, {}, {} and {}", key, keyId, url, organization);
-      // Read the key, keyId and url values
-      String keyValue = StringUtils.trimToNull((String) properties.get(key));
-      String keyIdValue = StringUtils.trimToNull((String) properties.get(keyId));
-      String urlValue = StringUtils.trimToNull((String) properties.get(url));
-      String organizationValue = StringUtils.trimToNull((String) properties.get(organization));
-
-      // Has the url signing provider been fully configured
-      // Note: organization is optional
-      if (keyValue == null || keyIdValue == null || urlValue == null) {
-        getLogger().debug(
-                "Unable to configure key with id '{}' and url matcher '{}' because the id, key or url is missing. Stopping to look for new keys.",
-                keyIdValue, urlValue);
-        break;
-      }
+    Enumeration<String> propertyKeys = properties.keys();
+    while (propertyKeys.hasMoreElements()) {
+      String propertyKey = propertyKeys.nextElement();
 
-      if (organizationValue == null) {
-        // organization is optimal. Set it to default
-        organizationValue = ANY_ORGANIZATION;
-        getLogger().debug("No organization set for key id {}, using default organization {}", keyIdValue,
-              organizationValue);
+      String[] parts = Arrays.stream(propertyKey.split("\\."))
+              .map(String::trim)
+              .toArray(String[]::new);
+      if (parts.length != 2) {
+        throw new ConfigurationException(propertyKey, "wrong property key format");
```

Comment:
That's nice.  Which key? :)

Reply:
Not sure what you are getting at here. You mean I should add which key was the culprit to the error message? Shouldn't the first parameter to `ConfigurationException` suffice? My hope was that the container displays that somewhere. :thinking:

Reply:
I believe (although I haven't actually tested) that the container will display the stack trace - `ConfigurationException` extends `RuntimeException`, but doesn't have any fancy logic saying to output the key name.  Generally I would drop a log entry at ERROR, then throw the `ConfigurationException`.

Reply:
There might not be specific logic, but `ConfigurationException` formats its message nicely. It looks like this:

```
2019-03-07 09:10:36,408 | ERROR | (Log:186) - [org.osgi.service.cm.ManagedService, org.opencastproject.security.urlsigning.provider.UrlSigningProvider, id=795, bundle=178/mvn:org.opencastproject/opencast-urlsigning-service-impl/7-SNAPSHOT]: Updating property key.foo.demoKeyOne.secret of configuration org.opencastproject.security.urlsigning.provider.impl.GenericUrlSigningProvider caused a problem: wrong property key format
```

followed by the stack trace, as you said.

I can add an additional log entry if you insist, of course, but I think that message contains everything we need to know, and I'm generally not a friend of excessive, unnecessary logging. :wink:

Reply:
I would argue that the average user isn't going to parse the whole message and will just complain on list, but technically you are correct (the best kind).